Fuzzing PreparedExpression#Execute and #ExecuteWithPositionalParams with columns and parameters

zd-project opened the pull request:
Added ParameterValueMap extractor to extract SQL statement variables from protobuf schema.

**There are two TODOs in this PR**
- [x] Handle the invalid numeric value 
- [x] Handle default literal value

zd-project committed:
Added ParameterValueMapExtractor

shumway commented on zetasql/fuzzing/protobuf/internal/parameter_value_map_extractor_test.cc:
```diff
@@ -0,0 +1,343 @@
+
+// TODO: Handle default value
+
+// template <typename ExprType>
```
Delete this commented-out code?

evmaus replied:
+1

evmaus commented on zetasql/fuzzing/protobuf/internal/parameter_value_map_extractor.h:
```diff
@@ -0,0 +1,52 @@
+
+class ParameterValueMapExtractor : public ProtoExprExtractor<zetasql::ParameterValueMap> {
+ public:
+
```
Maybe either comment these or remove the extra lines?

zd-project replied:
Might be a mistake. Will remove in next commit

evmaus commented on zetasql/fuzzing/protobuf/parameter_grammar.proto:
```diff
@@ -31,4 +33,45 @@ message Whitespace {
+
+// TODO: Support high precision value representation
+message NumericLiteral {
+    required bytes value = 1;
```
You may be able to alternatively get away with a series of int64s here instead of bytes, which would fix the problem with the bytes winding up being an invalid numeric value.

zd-project replied:
It does make literal conversion harder in that case, and we have to use zetasql::NumericValue to help us convert them to numeric literal. I would assume this is less ideal as we now are rely on components that we want to test, and can also limit our test input space (since the output is constrained by whatever the helper function from zetasql::NumericValue)

evmaus replied:
That makes sense.

evmaus reviewed (changes requested):
Mostly LGTM!

Just a few nits.  I'll do a final approval once you've finished the todos left in the PR.  :)

zd-project committed:
Introduce ArgumentExtractor for ParameterValueMap; Various Bug Fixes

Handled parameter_grammar::Default to zetasql::Value conversion
Handled invalid NumericLiteral to zetasql::Value conversion
Preliminary push on GetParamAs aliasing

zd-project commented:
Handled TODOs. Had another one specifically at [this line](https://github.com/googleinterns/zetasql-fuzzing/pull/10/commits/d0a15fa2e4624a67b1227f3de837988ff31abd0f#diff-6f421ef7348b2d0ae61d39972f0cb9c7R25). I would like to specialize and alias a function template which can correctly set the intent of an extracted ParameterValueMap argument to the fuzz target, but [the reference](https://en.cppreference.com/w/cpp/language/using_declaration#Notes) suggests it might be impossible. I'm wondering what are other elegant solutions we can take? @shumway @evmaus

zd-project committed:
Support fuzzing columns and parameters arguments in PreparedExpression target

evmaus reviewed (approved):
LGTM!

shumway commented on zetasql/fuzzing/component/fuzz_targets/prepared_expression_target.h:
```diff
@@ -24,10 +24,15 @@ namespace zetasql_fuzzer {
 
  private:
   std::unique_ptr<std::string> sql_expression;
+  std::unique_ptr<zetasql::ParameterValueMap> columns =
```
auto columns = std::make_unique...

Google-style says you should use auto here, since the type is declared when you use std::make_unique<T>.

shumway replied:
Nevermind, my mistake.

zd-project committed:
Separate literal_value_extractor

zd-project committed:
Fixed ParameterValueMapExtractor to extract intended Values

zd-project committed:
Introduced ParameterValueListExtractor; Supplemented ParameterValueMapExtractorTest

zd-project committed:
Created argument and extractor for positional parameters and fuzz target for ExecuteWithPositionalParams

zd-project commented:
I would probably suggest reading all commits changes because some of the implementation details, especially for ParameterValueMapExtractor and ParameterValueMapArg, have changed substantially. Additional testing has been introduced to ensure good functionality.

evmaus commented on zetasql/fuzzing/protobuf/internal/zetasql_expression_extractor_test.cc:
```diff
@@ -117,22 +120,33 @@ INSTANTIATE_TEST_SUITE_P(
-        std::make_tuple(ExtractableSpecialLiteral(LiteralExpr::V_NULL), "NULL")));
+    NullLiteralTest, ProtoExprExtractorTest,
+    ::testing::Combine(::testing::ValuesIn(ExtractableNullLiterals()), ::testing::Values("NULL")));
+    // ::testing::Values(
```
Maybe remove the commented code here?

evmaus reviewed (approved):
LGTM, not much to add other than some cleanup.
